### DISCUSSION TITLE
I can't send a notification to the users in a role

### DISCUSSION THREAD

**Discussion by elgransan** [oqtane/oqtane.framework#2229] [General]
Hi, I have a question regarding users...

I'm trying to send a notification from a user in a role to other users in another role, to ask them to complete a form, with this code
![image](https://user-images.githubusercontent.com/9949584/171275738-8ef7976f-64b9-403f-80d0-d777aa9e57f1.png)

But, I get an "Unauthorized UserRole Get Attempt For Site"

Which is the way to do this?

**Comment by sbwalker** [comment]
The UserRole GET API method requires that the user who is making a request to get the list of users in a specific role, is a member of that role themselves. This is to prevent the system from disclosing confidential information to unauthorized users. For example, a standard Registered User should not be able to call the API and determine the list of users who are Administrators or Hosts. 

This being said, the framework uses a least privilege philosophy ie. it locks down APIs by default and only provides access for specific use cases. In some cases this means it is more restrictive than it should be... and if a legitimate use case can be demonstrated where additional access should be granted, then it is worthy of consideration. Security is more of an art than a science - in many cases there is a subjective aspect to access restrictions which is based on personal perspective.

**Comment by elgransan** [comment]
Thank you @sbwalker I agree with you and you already told me that for another use case (which I changed the approach to follow the Oqtane security rules). But, I don't know how to do this in another way. I think it's a typical workflow case:

1. A user creates or modify an object
2. The object changes its state
3. The next role must be notified that the object was changed
4. The next role continue working with that object 

For now, I copied the controller's UserRule method into my module removing the constraint.

Maybe a method to send a notification to all the users in a role can solve that, and I don't even need to read all users

**Reply by sbwalker** [reply]
@elgransan you make a good point that in workflow scenarios a user will often need to assign an item to a more privileged role - a role which they are not a member of. And the current API method does not allow for this scenario. We could certainly relax the permissions on this method which would enable this use case, however it would have the side effect of allowing users to obtain the list of users in any role (note that the User object property values are trimmed based on additional security criteria - so information disclosure would still be mitigated). I believe that Active Directory allows a user to view the members of a group (unless the admin has explicitly denied this right) - so this change would be consistent with other systems.

**Reply by elgransan** [reply]
Excellent @sbwalker, let me know if I can help you doing this change. If you want, you can share me the interfaces that you think are necessaries/enough and I can do the changes, or I can make a full proposal.

**Reply by sbwalker** [reply]
#2237 should resolve the issue
